### PR TITLE
[yum] [dnf] add output of history info

### PR DESCRIPTION
**Pull request by didib**
Add a list of installed/erased/updated packages per yum/dnf transaction.

Signed-off-by: Yedidyah Bar David didi@redhat.com

### CODE REVIEW

**Comment by didib** [comment]
Tested with:

On RHEL7:
 sosreport -k yum.yum-history-info=on

On Fedora23:
 sosreport -k dnf.history-info=on -k dnf.history=on

Probably due to the way sosreport handles options, this didn't collect dnf history info:
 sosreport -k dnf.history-info=on

**Comment by didib** [comment]
@sandrobonazzola we might want these in ovirt-log-collector.

**Comment by sandrobonazzola** [comment]
:+1: on this change on my side. @didib once a release of sos with included this patch will be released it make sense to add the option to the calls to sos from ovirt-log-collector. I suggest to open a bz to track it if / when this patch will be accepted

**Comment by bmr-cymru** [comment]
Looks reasonable as a plugin option (although it may potentially be worth generalising in some way in the - future for other, non-RPM based package managers / depsolvers  maybe `--package-history` or something - no need to decide that now).

Patch looks fine at a quick glance too, I'll try to make time for a proper look later.

If you're interested in having this in any upcoming Red Hat releases then it is definitely time to file bugzillas - we just released 3.3 a few weeks ago as the planned rebase for upcoming RHEL updates.

**Comment by didib** [comment]
Now opened these:
https://bugzilla.redhat.com/show_bug.cgi?id=1360621 oVirt
https://bugzilla.redhat.com/show_bug.cgi?id=1360623 Fedora
https://bugzilla.redhat.com/show_bug.cgi?id=1360624 RHEL7

File: sos/plugins/dnf.py [hunk 1]
```diff
@@ -60,4 +66,17 @@ def setup(self):
         if self.get_option("history"):
             self.add_cmd_output("dnf history")
 
+        if self.get_option("history-info"):
+            history = self.call_ext_prog("dnf history")
```
Comment (didib) [inline sos/plugins/dnf.py]: Isn't that what lines 66-67 already do?

If you mean "do that even if user didn't pass the option 'history'", then I might agree, but:
1. Will let sos maintainer decide, not sure what's the best way to do that. I don't mind pushing an updated patch if needed.
2. It seems to already be how it works, see my previous comment - '-k dnf.history-info=on' didn't work for me, only '-k dnf.history-info=on -k dnf.history=on' did. Not sure that's intentional, or a by-product of how options are handled.

Comment (pmoravec) [inline sos/plugins/dnf.py]: Ah, I was partially blind.. Doesn't it makes sense to implicitly enable `history` option when `history-info` is enabled?

Comment (didib) [inline sos/plugins/dnf.py]: As I wrote, perhaps it does - not sure if that's considered ok here. @bmr-cymru what do you say?

Comment (bmr-cymru) [inline sos/plugins/dnf.py]: > Ah, I was partially blind.. Doesn't it makes sense to implicitly enable history option when history-info is enabled?

This is how we've handled previous cases (e.g. `-k rpm.rpmva` enables `--verify` just for the RPM module).

But.. I don't really like the proliferation of options here - experience shows that users don't tend to find them unless pointed to them explicitly (which somewhat removes the point of having them - if a 2nd touch is needed to re-request sos data with `-k dnf.history*` then we almost might just as well ask for `dnf history-whatever` itself..).

What are the actual costs for these operations, and can they not be made default (at least in the case of `dnf history`)?

Comment (didib) [inline sos/plugins/dnf.py]: Ran each of these 3 times, on two machines - rhel7 and fedora24. Time is the best of 3. The fedora machine has 57 entries in dnf history, the rhel7 one has 28 in yum history.

time sosreport --batch
fedora24 22s
rhel7 42s

time sosreport -k yum.yum-history-info=on --batch
rhel7 52s

time sosreport -k dnf.history-info=on -k dnf.history=on --batch
fedora24 44s

So around 10 seconds more on the rhel7 machine, 22 seconds more on the fedora24 one. Since it seems somewhat proportional to the number of transactions, I'd say it adds around half a second per transaction. Very unscientific, I know.

Didn't try just adding yum/dnf history (without info per transaction), but when running them manually they take around 0.3s on rhel7, 0.6s on fedora24 (or probably around 0.01 seconds per entry).

Also note that at least on my rhel7 machine, the output of 'yum history' is cut in the middle:
...
    11 | downgrade http://bob.eng | 2016-06-14 11:49 | Downgrade      |    1   
    10 | install http://bob.eng.l | 2016-06-14 11:43 | Update         |    1   
     9 |                          | 2016-06-14 11:33 | I, U           |   28 EE
history list

but e.g. 'yum history info 7' does show info. Might be a problem in yum, but I'd rather collect all the info anyway.

So bottom line, I'd say the time added is significant, but IMO acceptable.

**Comment by sandrobonazzola** [comment]
@bmr-cymru anything else needed here?

**Comment by bmr-cymru** [comment]
Nope - all good I think - I just need a day to catch up on pulls at the moment.

File: sos/plugins/dnf.py (hunk 1 above)
Comment (didib) [inline sos/plugins/dnf.py]: > But.. I don't really like the proliferation of options here - experience shows that users don't tend to find them unless pointed to them explicitly (which somewhat removes the point of having them - if a 2nd touch is needed to re-request sos data with -k dnf.history* then we almost might just as well ask for dnf history-whatever itself..).

Sorry I didn't reply to this point earlier. My current incentive is ovirt-log-collector, which calls sosreport, and will then be able to pass these options as needed, when available. For general/default use of sosreport, not sure. I think there is value in adding them always by default, as they do not take an unreasonable amount of time and quite little space. But I am not sure I am in a position to decide - it's the maintainer who knows what kinds of requests/complaints current users have, and might be able to better expect whether adding this amount of time will increase the "sosreport is too slow" complaints more than the win in less "I already sent a sosreport, do I really need to run it again? Can't you make it get you all you need at once?" kind of complaints, or the other way around...

**Comment by sandrobonazzola** [comment]
@bmr-cymru take your time, we're not in a hurry (unless you release next sos version tomorrow :-) )

**Comment by bmr-cymru** [comment]
:+1:

**Comment by sandrobonazzola** [comment]
@bmr-cymru 3 months passed, can you please look into this?

**Comment by bmr-cymru** [comment]
I've made a small change to make the `dnf` plugin option tuple fit on a single line (it's strongly preferred).

Generally it seems a bit funny to add the `yum` plugin `history-info` option, but to not add the `history` option (that `dnf` has had for ages), but I won't hold this patch up any longer to add it - `yum` is legacy now as far as `sos` is concerned and it is unlikely at this point that RHEL6 or older will ever ship `3.3` or later.
